src/igraph_eulerian.c

```diff
@@ -0,0 +1,652 @@
+                break;
+            }
+        }
+
+    /* strong connectivity is NOT required for just a path */
+    } else if ((outgoing_excess == 1 && incoming_excess == 1) && (res_weak)) {
+        *has_path = 1;
+        *has_cycle = 0;
+    } else {
+        *has_path = 0;
+        *has_cycle = 0;
+    }
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &il, IGRAPH_OUT));
+    IGRAPH_FINALLY(igraph_inclist_destroy, &il);
+
+    /* check if singleton self loops destroy the Eulerian path/cycle */
+    vsize = igraph_vector_size(&check_for_self_loops);
+    for (i = 0; i < vsize; i++) {
+        igraph_vector_int_t *incedges;
+        incedges = igraph_inclist_get(&il, VECTOR(check_for_self_loops)[i]);
+        long int nc = igraph_vector_int_size(incedges);
+        if (nc > 0) {
+            *has_path = 0;
+            *has_cycle = 0;
+        }
+    }
+
+    igraph_vector_destroy(&check_for_self_loops);
+    igraph_inclist_destroy(&il);
+    igraph_vector_destroy(&csize_weak);
+    igraph_vector_destroy(&in_degree);
+    igraph_vector_destroy(&out_degree);
+    igraph_vector_destroy(&cluster_member);
+    IGRAPH_FINALLY_CLEAN(6);
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/**
+ * \ingroup Eulerian
+ * \function igraph_is_eulerian
+ * \brief Checks whether an Eulerian path or cycle exists
+ *
+ * An Eulerian path traverses each edge of the graph precisely once. A closed
+ * Eulerian path is referred to as an Eulerian cycle.
+ *
+ * \param graph The graph object.
+ * \param has_path Pointer to a Boolean, will be set to true if an Eulerian path exists.
+ * \param has_cycle Pointer to a Boolean, will be set to true if an Eulerian cycle exists.
+ * \return Error code:
+ *         \c IGRAPH_ENOMEM, not enough memory for
+ *         temporary data.
+ *
+ * Time complexity: O(|V|+|E|), the
+ * number of vertices times the number of edges.
+ *
+ */
+
+int igraph_is_eulerian(const igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle) {
+    igraph_integer_t start_of_path = 0;
+    
+    if (igraph_is_directed(graph)) {
+        IGRAPH_CHECK(igraph_i_is_eulerian_directed(graph, has_path, has_cycle, &start_of_path));
+    } else {
+        IGRAPH_CHECK(igraph_i_is_eulerian_undirected(graph, has_path, has_cycle, &start_of_path));
+    } 
+    return IGRAPH_SUCCESS;
+}
+
+
+static int igraph_i_eulerian_path_undirected(const igraph_t *graph, igraph_vector_t *edge_res, igraph_vector_t *vector_res, igraph_integer_t start_of_path) {
+    long int curr;
+    igraph_integer_t n, m;
+    igraph_inclist_t il;
+    igraph_stack_t path, tracker, edge_tracker, edge_path;
+    igraph_vector_bool_t visited_list;
+    igraph_vector_t degree;
+
+    n = igraph_vcount(graph);
+    m = igraph_ecount(graph);
+
+    if (edge_res) {
+        igraph_vector_clear(edge_res);
+        IGRAPH_CHECK(igraph_vector_reserve(edge_res, m));
+    }
+
+    if (vector_res) {
+        igraph_vector_clear(vector_res);
+        IGRAPH_CHECK(igraph_vector_reserve(vector_res, n+1)); 
```

Comment:
I believe this should be `m+1`, not `n+1`. The final size of this vector will be `m+1`.

Reply:
Addressed in 2d36c3c